### ISSUE TITLE
ner_few_shot_ru  | Fine-tuning the model

### ISSUE THREAD

**Issue by vcjob** [deepmipt/DeepPavlov#1071]
Hello everyone! 

I read the documentation at http://docs.deeppavlov.ai/en/master/features/models/ner.html#few-shot-language-model-based and it says, that I need to tag at least 10 sentences to start training. So with all the info from the page in mind, I came up with the following questions: 
1. If we have 10 sentences with entity marks, what about the number of the ones without? also 10? 10k or maybe a couple of millions? Should they just be in format "word per line, empty line between sentences"?
2. In that case, the SVM is used.  Is it trained on those 10 sentences and used afterwards to classify the rest sentences? What is the "K" number?
3. Is it fine-tuning the BERT model, or training from scratch with the help of LM?
4. It says also, that if there are more than 1k sentences available, better to use ner_few_shot_ru_simulate. Nevertheless, if execute the code from the same documentary page:
_from deeppavlov import configs, train_model
ner_model = train_model(configs.ner.ner_few_shot_ru, download=True)_
it downloads a huge training, testing and validation sets and starts to train the model (moreover, on the cpu). Have I misunderstood anything?
5. May I use my own tags, like "DOCS"?
6. Still, how many sentences are recommended to fine-tune the model with addition of new tags; How many sentences do I need to train the model from scratch? 

Thank you a lot for the answers! Looking forward to them!

**Comment by yoptar** [comment]
Hi @vcjob,

1. it would be better if distribution of sentences that have entities and not is about the same in training data as in real live
2. so if you have no tagged data at all, you start to mark up your raw data until you have at least K (for example 10) sentences with entities of interest. But you also use the sentences without tags in training.
3. It's training from scratch, yes.
4. I just tried to execute
    ```python3
    from deeppavlov import configs, train_model
    ner_model = train_model(configs.ner.ner_few_shot_ru, download=True)
    ```
    It downloaded the language model and raised an error `train.txt not found in "/data/deeppavlov_data/downloads/ner_few_shot_data"` as it should
5. Sure. The model itself works with tag indexes in a [vocab](https://github.com/deepmipt/DeepPavlov/blob/master/deeppavlov/configs/ner/ner_few_shot_ru.json#L20-L27) that is filled at the training stage
6. It depends on the data and desired quality. You can start at 10, check on other data and just use it if you're satisfied or mark up and add more training data if you're not. At some point you can say that it will never be good enough and switch to a more complex model that requires even more data, but this few shot model is usually a good start.

**Comment by vcjob** [comment]
> Hi @vcjob,
> 
> 1. it would be better if distribution of sentences that have entities and not is about the same in training data as in real live
> 2. so if you have no tagged data at all, you start to mark up your raw data until you have at least K (for example 10) sentences with entities of interest. But you also use the sentences without tags in training.
> 3. It's training from scratch, yes.
> 4. I just tried to execute
>    ```python
>    from deeppavlov import configs, train_model
>    ner_model = train_model(configs.ner.ner_few_shot_ru, download=True)
>    ```
>    
>    
>    It downloaded the language model and raised an error `train.txt not found in "/data/deeppavlov_data/downloads/ner_few_shot_data"` as it should
> 5. Sure. The model itself works with tag indexes in a [vocab](https://github.com/deepmipt/DeepPavlov/blob/master/deeppavlov/configs/ner/ner_few_shot_ru.json#L20-L27) that is filled at the training stage
> 6. It depends on the data and desired quality. You can start at 10, check on other data and just use it if you're satisfied or mark up and add more training data if you're not. At some point you can say that it will never be good enough and switch to a more complex model that requires even more data, but this few shot model is usually a good start.

Thank you a lot for the reply!
Would you be also so kind to suggest, having 10k sentences dataset for NER at my disposal, should I try using _ner_few_shot_ru_simulate_ to train NER model, or maybe I should better try to fine-tune BERT (if so, extra question: should I train from scratch of fine-tune existing DP BERT model, and if so, is there any guide of instructions on how to do it?).
The number of tags is around 10. 

Any suggestions are highly appreciated!

**Comment by yoptar** [comment]
> should I try using ner_few_shot_ru_simulate to train NER model from scratch
This config will use only a few samples from a big dataset to show how the few-shot model is trained, so no.

I would advise you to use one of our BERT-based configs \([ner_ontonotes_bert_mult](https://github.com/deepmipt/DeepPavlov/blob/master/deeppavlov/configs/ner/ner_ontonotes_bert_mult.json) on multi-language BERT, [ner_ontonotes_bert](https://github.com/deepmipt/DeepPavlov/blob/master/deeppavlov/configs/ner/ner_ontonotes_bert.json) for English or [ner_rus_bert](https://github.com/deepmipt/DeepPavlov/blob/master/deeppavlov/configs/ner/ner_rus_bert.json) for Russian\). Multi-language model can show the best results and might even work on languages it was not trained on, but it's heavier.

The code for training can look like this:
```python3
import json
from deeppavlov import configs, build_model, train_model

with configs.ner.ner_ontonotes_bert_mult.open(encoding='utf8') as f:
    ner_config = json.load(f)

ner_config['dataset_reader']['data_path'] = '~/my_data_dir/'  # directory with train.txt, valid.txt and test.txt files
ner_config['metadata']['variables']['NER_PATH'] = '~/where_to_save_the_model/'
ner_config['metadata']['download'] = [ner_config['metadata']['download'][-1]]  # do not download the pretrained ontonotes model

ner_model = train_model(ner_config, download=True)
```

**Comment by vcjob** [comment]
> > should I try using ner_few_shot_ru_simulate to train NER model from scratch
> 
> This config will use only a few samples from a big dataset to show how the few-shot model is trained, so no.
> 
> I would advise you to use one of our BERT-based configs ([ner_ontonotes_bert_mult](https://github.com/deepmipt/DeepPavlov/blob/master/deeppavlov/configs/ner/ner_ontonotes_bert_mult.json) on multi-language BERT, [ner_ontonotes_bert](https://github.com/deepmipt/DeepPavlov/blob/master/deeppavlov/configs/ner/ner_ontonotes_bert.json) for English or [ner_rus_bert](https://github.com/deepmipt/DeepPavlov/blob/master/deeppavlov/configs/ner/ner_rus_bert.json) for Russian). Multi-language model can show the best results and might even work on languages it was not trained on, but it's heavier.
> 
> The code for training can look like this:
> 
> ```python
> import json
> from deeppavlov import configs, build_model, train_model
> 
> with configs.ner.ner_ontonotes_bert_mult.open(encoding='utf8') as f:
>     ner_config = json.load(f)
> 
> ner_config['dataset_reader']['data_path'] = '~/my_data_dir/'  # directory with train.txt, valid.txt and test.txt files
> ner_config['metadata']['variables']['NER_PATH'] = '~/where_to_save_the_model/'
> ner_config['metadata']['download'] = [ner_config['metadata']['download'][-1]]  # do not download the pretrained ontonotes model
> 
> ner_model = train_model(ner_config, download=True)
> ```

Thank you a lot!
I will try soon.
Have a great day!

**Comment by xirect** [comment]
If you use the multilingual bert model as a base and train it on one language, will it use zero-shot transfer then too to teach the other languages the new tags?

**Comment by paulagd** [comment]
> > should I try using ner_few_shot_ru_simulate to train NER model from scratch
> 
> This config will use only a few samples from a big dataset to show how the few-shot model is trained, so no.
> 
> I would advise you to use one of our BERT-based configs ([ner_ontonotes_bert_mult](https://github.com/deepmipt/DeepPavlov/blob/master/deeppavlov/configs/ner/ner_ontonotes_bert_mult.json) on multi-language BERT, [ner_ontonotes_bert](https://github.com/deepmipt/DeepPavlov/blob/master/deeppavlov/configs/ner/ner_ontonotes_bert.json) for English or [ner_rus_bert](https://github.com/deepmipt/DeepPavlov/blob/master/deeppavlov/configs/ner/ner_rus_bert.json) for Russian). Multi-language model can show the best results and might even work on languages it was not trained on, but it's heavier.
> 
> The code for training can look like this:
> 
> ```python
> import json
> from deeppavlov import configs, build_model, train_model
> 
> with configs.ner.ner_ontonotes_bert_mult.open(encoding='utf8') as f:
>     ner_config = json.load(f)
> 
> ner_config['dataset_reader']['data_path'] = '~/my_data_dir/'  # directory with train.txt, valid.txt and test.txt files
> ner_config['metadata']['variables']['NER_PATH'] = '~/where_to_save_the_model/'
> ner_config['metadata']['download'] = [ner_config['metadata']['download'][-1]]  # do not download the pretrained ontonotes model
> 
> ner_model = train_model(ner_config, download=True)
> ```

Hi, I have followed your discussion. So, if I would like to use **ner_ontonotes_bert_mult** model but to detect my own tags (lets put TAG1, TAG2). Is it enought to put my data in **'~/my_data_dir/'** path and  use the ner_model you define in this piece of code?

Also, in this case the model will detect just my two tags or also the ones defined in the documentation for this model?

Thank you a lot in advance!

**Comment by yoptar** [comment]
Hi @xirect,

> If you use the multilingual bert model as a base and train it on one language, will it use zero-shot transfer then too to teach the other languages the new tags?

This is exactly how we got our [multilingual NER model](https://demo.deeppavlov.ai/#/mu/ner), it was trained only on the English dataset.

**Comment by yoptar** [comment]
Hi @paulagd,

> So, if I would like to use ner_ontonotes_bert_mult model but to detect my own tags (lets put TAG1, TAG2). Is it enought to put my data in '~/my_data_dir/' path and use the ner_model you define in this piece of code?

Yes. And after training it once you can run `model = build_model(ner_config)` to initialize again. I also suggest save the updated configuration with new paths to another file so you can reuse it easier.

> Also, in this case the model will detect just my two tags or also the ones defined in the documentation for this model?

It will only detect your tags. You can try and merge your data with the provided [conll2003 dataset](http://files.deeppavlov.ai/deeppavlov_data/conll2003_v2.tar.gz) to train the model on everything, but the quality will most likely drop as your data has no conll2003 labels and conll2003 data does not have yours.

**Comment by yoptar** [comment]
I'm closing this issue as all of the questions are answered. If you have any followup questions, please ask them on [our forum](https://forum.deeppavlov.ai/).
We're trying to optimize our communication channels and leave issues for code issues and requests only and use the forum for general questions.

**Comment by paulagd** [comment]
Okey, following discussion in [here](https://forum.deeppavlov.ai/t/error-in-training-multilingual-ner-with-own-data/452). 

Thanks!

**Comment by paulagd** [comment]
> I'm closing this issue as all of the questions are answered. If you have any followup questions, please ask them on [our forum](https://forum.deeppavlov.ai/).
> We're trying to optimize our communication channels and leave issues for code issues and requests only and use the forum for general questions.

Okey, following discussion in [here](https://forum.deeppavlov.ai/t/error-in-training-multilingual-ner-with-own-data/452). 

Thanks!
